### ISSUE TITLE
`frizbee actions` generates wrong/invalid commit ref?

### ISSUE THREAD

**Issue by woodruffw** [stacklok/frizbee#206]
### Describe the issue

Hi there! Thanks for `frizbee`, it's an awesome tool.

@chenrui333 ran `frizbee` on `zizmor` here: https://github.com/woodruffw/zizmor/pull/90 and it produced a strange (seemingly nonexistent) pinned commit ref.

In particular, this:

```yaml
      - name: Install the latest version of uv
        uses: astral-sh/setup-uv@v3
```

became this:

```yaml
      - name: Install the latest version of uv
        uses: astral-sh/setup-uv@c5df01c9e4233fcd789bd004850aa36809d07727 # v3
```

However, that commit (`c5df01c9e4233fcd789bd004850aa36809d07727`) doesn't appear to exist at all, either on the upstream or anywhere in the fork network: <https://github.com/astral-sh/setup-uv/commit/c5df01c9e4233fcd789bd004850aa36809d07727>

I _think_ the correct ref here should be `3b9817b1bf26186f03ab8277bab9b827ea5cc254`, corresponding to tags `v3` and `v3.2.0`: https://github.com/astral-sh/setup-uv/releases/tag/v3.2.0

### To Reproduce

To reproduce, run `frizbee actions .github/workflows` on https://github.com/woodruffw/zizmor and observe the ref 🙂 

`frizbee version` output:

```
Version: 0.1.4
Go Version: go1.23.2
Git Commit: 
Commit Date: 
OS/Arch: darwin/arm64
Dirty: false
```

### What version are you using?

v0.1.4

**Comment by woodruffw** [comment]
Minimal reproducer:

```
$ frizbee actions astral-sh/setup-uv@v3
astral-sh/setup-uv@c5df01c9e4233fcd789bd004850aa36809d07727
```

**Comment by woodruffw** [comment]
GitHub's own API endpoints appear to be returning the wrong ref here:

```
http GET https://api.github.com/repos/astral-sh/setup-uv/git/refs/tags/v3
```

yields:

```json
{
    "node_id": "REF_kwDOMnghHqxyZWZzL3RhZ3MvdjM",
    "object": {
        "sha": "c5df01c9e4233fcd789bd004850aa36809d07727",
        "type": "tag",
        "url": "https://api.github.com/repos/astral-sh/setup-uv/git/tags/c5df01c9e4233fcd789bd004850aa36809d07727"
    },
    "ref": "refs/tags/v3",
    "url": "https://api.github.com/repos/astral-sh/setup-uv/git/refs/tags/v3"
}
```

...and `object.url` points to the real commit ref seemingly:

```
http GET https://api.github.com/repos/astral-sh/setup-uv/git/tags/c5df01c9e4233fcd789bd004850aa36809d07727
```

yields:

```json
{
  "node_id": "TA_kwDOMnghHtoAKGM1ZGYwMWM5ZTQyMzNmY2Q3ODliZDAwNDg1MGFhMzY4MDlkMDc3Mjc",
  "sha": "c5df01c9e4233fcd789bd004850aa36809d07727",
  "url": "https://api.github.com/repos/astral-sh/setup-uv/git/tags/c5df01c9e4233fcd789bd004850aa36809d07727",
  "tagger": {
    "name": "eifinger",
    "email": "eifinger@users.noreply.github.com",
    "date": "2024-10-25T12:13:43Z"
  },
  "object": {
    "sha": "3b9817b1bf26186f03ab8277bab9b827ea5cc254",
    "type": "commit",
    "url": "https://api.github.com/repos/astral-sh/setup-uv/git/commits/3b9817b1bf26186f03ab8277bab9b827ea5cc254"
  },
  "tag": "v3",
  "message": "Release v3.2.0\n",
  "verification": {
    "verified": false,
    "reason": "unsigned",
    "signature": null,
    "payload": null
  }
}
```

**Comment by blkt** [comment]
@woodruffw it looks like there are two steps involved
* first obtain the tag's identifier using its name
```sh
$ gh api /repos/astral-sh/setup-uv/git/refs/tags/v3
{
  "ref": "refs/tags/v3",
  "node_id": "REF_kwDOMnghHqxyZWZzL3RhZ3MvdjM",
  "url": "https://api.github.com/repos/astral-sh/setup-uv/git/refs/tags/v3",
  "object": {
    "sha": "57b15becf94238836cd31b932339e7d72a15ef9c",
    "type": "tag",
    "url": "https://api.github.com/repos/astral-sh/setup-uv/git/tags/57b15becf94238836cd31b932339e7d72a15>
  }
}
```
* then use `object.sha` to get the details of the tag
```sh
$ gh api /repos/astral-sh/setup-uv/git/tags/57b15becf94238836cd31b932339e7d72a15ef9c
{
  "node_id": "TA_kwDOMnghHtoAKDU3YjE1YmVjZjk0MjM4ODM2Y2QzMWI5MzIzMzllN2Q3MmExNWVmOWM",
  "sha": "57b15becf94238836cd31b932339e7d72a15ef9c",
  "url": "https://api.github.com/repos/astral-sh/setup-uv/git/tags/57b15becf94238836cd31b932339e7d72a15ef>
  "tagger": {
    "name": "eifinger",
    "email": "eifinger@users.noreply.github.com",
    "date": "2024-11-15T08:23:29Z"
  },
  "object": {
    **"sha": "e779db74266a80753577425b0f4ee823649f251d",**
    "type": "commit",
    "url": "https://api.github.com/repos/astral-sh/setup-uv/git/commits/e779db74266a80753577425b0f4ee8236>
  },
  "tag": "v3",
  "message": "Release v3.2.3\n",
  "verification": {
    "verified": false,
    "reason": "unsigned",
    "signature": null,
    "payload": null,
    "verified_at": null
  }
}
```
* finally, get `object.sha` again to get the actual commit (`type` equals `commit`).

I will open a PR shortly, but I'd love to hear opinions from @jhrozek and @JAORMX, plus some better testing before we can merge it.

**Comment by woodruffw** [comment]
Thanks @blkt! Those steps looks right to me, and I greatly appreciate your response!

**Comment by JAORMX** [comment]
I'm still not quite sure why we're getting this unexistent ref to begin with. Have we dug into that?

**Comment by woodruffw** [comment]
> I'm still not quite sure why we're getting this unexistent ref to begin with. Have we dug into that?

The ref does exist: it's just a tag ref, not a commit ref. This ends up working since GitHub knows how to handle both, but GitHub generally only shows commit refs in the UI itself (including on the tags/releases pages).

TL;DR both refs are correct, but the commit ref is the "canonical" one as far as GitHub is concerned.

**Comment by JAORMX** [comment]
Ah! That makes sense. And I see @blkt even has a patch in the works

**Comment by blkt** [comment]
The main difference is that "partial" tags (e.g. `v3` when you have `v3.6.0`) resolve to a tag object, while "full" tags (e.g. the `v3.6.0`) resolve to commits objects directly, so it worked depending on the exact tagging mechanism.

**Comment by woodruffw** [comment]
> The main difference is that "partial" tags (e.g. `v3` when you have `v3.6.0`) resolve to a tag object, while "full" tags (e.g. the `v3.6.0`) resolve to commits objects directly, so it worked depending on the exact tagging mechanism.

I could be wrong, but I think the reason underneath this is convention: the "partial" tags *could* resolve to commit objects if they were re-tagged using the commit instead of the "full" tag, but in practice GitHub's official actions (and many others) seem to do the former.

(The interactions here are very confusing: I think the behavior also varies if the tag is a "plain" tag versus an "annotated" one.)

**Comment by woodruffw** [comment]
@JAORMX Thanks for the fix here! Would it be possible to get this landed in a release? Right now `brew` et al. are carrying patches for this, which is not ideal.

**Comment by JAORMX** [comment]
@woodruffw done did https://github.com/stacklok/frizbee/releases/tag/v0.1.5

**Comment by woodruffw** [comment]
Thank you!
